### ISSUE TITLE
Silent: Remove verbose error if python not loadable 

### ISSUE THREAD

**Issue by tinmarino** [SirVer/ultisnips#1237]
Hi, awesome plugin, I `:packadd` it in my vimrc.

__Problem:__ But on many machines, vim python is not well instaled (my bad).
__Solution:__ It would be great if Ultisnip just silently fails in that case.
__My problem, my job:__ I can do a PR, now I'm asking if dev agrees with this __silent fail__.

Thanks guys.

PS : the message:
```text
Error detected while processing /home/almamgr/.vim/pack/bundle/opt/ultisnips/autoload/UltiS
nips.vim:
line    7:
E370: Could not load library /usr/lib/python3.8/config-3.8-x86_64-linux-gnu/libpython3.8.so

Error detected while processing /home/almamgr/.vim/pack/bundle/opt/ultisnips/autoload/UltiS
nips.vim:
line    7:
E263: Sorry, this command is disabled, the Python library could not be loaded.
Error detected while processing /home/almamgr/.vim/pack/bundle/opt/ultisnips/autoload/UltiS
nips.vim:
line    8:
E370: Could not load library /usr/lib/python3.8/config-3.8-x86_64-linux-gnu/libpython3.8.so

Error detected while processing /home/almamgr/.vim/pack/bundle/opt/ultisnips/autoload/UltiS
nips.vim:
line    8:
E263: Sorry, this command is disabled, the Python library could not be loaded.
Error detected while processing function UltiSnips#TrackChange:
line    1:
E370: Could not load library /usr/lib/python3.8/config-3.8-x86_64-linux-gnu/libpython3.8.so

Error detected while processing function UltiSnips#TrackChange:
line    1:
E263: Sorry, this command is disabled, the Python library could not be loaded.
```

**Comment by lacygoill** [comment]
I'm not a dev, but I agree that the stack trace is annoying.

FWIW, I would prefer that – on startup – UltiSnips tests whether the python interface is available, and if not bail out, and log one message so that the user has a chance to understand why the plugin doesn't work by reviewing the output of `:messages`.

**Comment by henk84** [comment]
I am getting a very similar behaviour:
upon entering insert mode, I get the following message on machines where no python3 is installed:
```
Error detected while processing /root/.local/share/nvim/plugged/ultisnips/autoload/UltiSnips.vim:                                                                             
line   10:                                                                                                                                                                    
E319: No "python3" provider found. Run ":checkhealth provider"                                                                                                                
```

That message is repeated on every keystroke which is very annoying.
I’m assuming this is the same or at least a closely related issue. Please let me know if I’m wrong and/or should make a separate issue.

My workaround is wrapping the call to load the plugin into a conditional:
```
if has("python3")
       Plug 'https://github.com/SirVer/ultisnips'
endif
```

**Comment by SirVer** [comment]
@henk84 I think your solution is not a workaround, it is the actual correct way of solving this issue. After all, you should not install and use a plugin if you do not have it's requirement (which is python) available.

**Comment by henk84** [comment]
> @henk84 I think your solution is not a workaround, it is the actual correct way of solving this issue. After all, you should not install and use a plugin if you do not have it's requirement (which is python) available.

TLDR: I agree with what @lacygoill says, more detailed explanation follows:
> FWIW, I would prefer that – on startup – UltiSnips tests whether the python interface is available, and if not bail out, and log one message so that the user has a chance to understand why the plugin doesn't work by reviewing the output of `:messages`.

I have to say that I’m a bit torn what the correct solution is. I think there are several facets to this issue that should be taken into account which have to do with different ways how configs can be written.
My particular case is that I handle most configs in my home using https://github.com/RichiH/vcsh and use that to deploy them to every host I use on a regular basis.
The installed software on those systems is not always under my control, meaning in particular that I’m not always in control of whether dependencies of e.g. vim-plugins I use are installed.
But even in cases where I have this kind of control, it’s not always a given that I want them to be installed, e.g. in the case of less powerful hardware, like single-board computers for a very narrow and specific purpose, where I want to keep the number of installed packages to a minimum.

There are several possible approaches to how to deal with this that come to mind.
1) One of them is to make sure I deploy exactly the config variant to each machine that works on this machine. This puts a lot of burden on the management of the config files in question. E.g. in this case I should not deploy a config trying to load this plugin to machines where I don’t want to have the dependencies installed.
This is something I don’t really want to do because it adds a lot of variations to a lot of configs that all need to be maintained.

2) The next would be what I did a few comments up: add safeguards around all config statements that could cause this kind of trouble.
This is also something I don’t really want to do because it can really blow up the configs in some cases. I can’t really make a good estimation but some plugins have a lot of dependencies. Also some people load a lot of plugins and trippling each call in question (surrounding the `Plug …` with `if` and `endif`) or even quintupling (adding an `else` statement notifying the user) seems very verbose and quickly makes the config unwieldy. In some cases one may have to check for multiple dependencies which could make one single `Plug …` blow up to 10 lines, I think. Maybe I’m missing some way to make this smaller but I don’t really like it in principle.

AFAICT so far what seems to be the best solution to this particular problem can be boiled down to "fail gracefully".
I would describle the current behaviour as "fail annoyingly": the message about the problem being repeated with almost every keystroke is very annoying and makes it very hard to use vim under these conditions und IMHO this should not happen.
"Failing gracefully" or "being fault tolerant" would mean exactly what @lacygoill described: one message on startup that the plugin could not be initialized, ideally with the reason being given, and then just be silent about it.

@SirVer do you think it would be possible to handle this situation like that?

**Comment by SirVer** [comment]
@henk84 I still remain that the right solution is to exclude the plugin in your config conditionally. I personally also would be as annoyed by a one time error on startup than a continious one, but I do resonate that "fail gracefully" is preferable for some people than "fail annoyingly" (which I prefer, because it increases pressure for a proper fix).

So, If a patch is proposed that fails gracefully, I am more than happy to review and consider it.

**Comment by SirVer** [comment]
I have implemented the suggested fix now in #1658 ; an error message is shown if python is not available, but only once, not every time. The error message can also be suppressed by a (undocumented) variable.
